### PR TITLE
Initial implementation of Stateful Precompile Contracts

### PR DESCRIPTION
**Pull request by evgeniy-scherbina**
Related document: https://kava-labs.atlassian.net/wiki/spaces/ENG/pages/1385136133/Go-Ethereum+Source+Code+Changes

### CODE REVIEW

File: contracts/contracts/ExampleSum3.sol
Comment (drklee3) [inline contracts/contracts/ExampleSum3.sol:1]: should we also include calls with the other opcodes that has the precompile changes so we can also test those are working? StaticCall, DelegateCall, CallCode

File: precompile/contract/contract.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+// NewStatefulPrecompileContract generates new StatefulPrecompile using [functions] as the available functions and [fallback]
+// as an optional fallback if there is no input data. Note: the selector of [fallback] will be ignored, so it is required to be left empty.
+func NewStatefulPrecompileContract(functions []*StatefulPrecompileFunction) (StatefulPrecompiledContract, error) {
+	// Construct the contract and populate [functions].
+	contract := &statefulPrecompileWithFunctionSelectors{
+		functions: make(map[string]*StatefulPrecompileFunction),
+	}
+	for _, function := range functions {
+		_, exists := contract.functions[string(function.selector)]
```
Comment (drklee3) [inline precompile/contract/contract.go:60]: Maybe also check and return error if selector is empty or not matching `SelectorLen`? And include a test with it error when trying to use a fallback function.

I think fine to exclude fallback function for now until we need it later, but would be nice to enforce it with an error until then

File: precompile/contract/utils.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+func DeductGas(suppliedGas uint64, requiredGas uint64) (uint64, error) {
+	if suppliedGas < requiredGas {
+		return 0, fmt.Errorf("out of gas, supplied: %v, required: %v", suppliedGas, requiredGas)
+	}
+	return suppliedGas - requiredGas, nil
+}
+
+// ParseABI parses the given ABI string and returns the parsed ABI.
+// If the ABI is invalid, it panics.
+func ParseABI(rawABI string) abi.ABI {
```
Comment (drklee3) [inline precompile/contract/utils.go]: Suggestion: Rename to `MustParseABI` to match similar naming of the other functions that panic if there's an error

File: contracts/contracts/ExampleSum3.sol
Comment (evgeniy-scherbina) [reply contracts/contracts/ExampleSum3.sol:1]: good point, I'll add e2e tests for those calls

File: precompile/contract/utils.go (hunk 2 above)
Comment (evgeniy-scherbina) [reply precompile/contract/utils.go]: agree

File: precompile/contract/contract.go (hunk 1 above)
Comment (evgeniy-scherbina) [reply precompile/contract/contract.go:60]: make sense, will do

Comment (nddeluca) [reply precompile/contract/contract.go:60]: Agree @drklee3, likely extend to also support receive in order to separate receive & fallback functionality of solidity interfaces

File: precompile/contract/contract.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	accessibleState AccessibleState,
+	caller common.Address,
+	addr common.Address,
+	input []byte,
+	suppliedGas uint64,
+	readOnly bool,
+) (ret []byte, remainingGas uint64, err error) {
+	// Otherwise, an unexpected input size will result in an error.
+	if len(input) < SelectorLen {
+		return nil, suppliedGas, fmt.Errorf("missing function selector to precompile - input length (%d)", len(input))
```
Comment (nddeluca) [inline precompile/contract/contract.go:82]: I don't believe these error cases are tested

File: precompile/contract/utils.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+func MustCalculateFunctionSelector(functionSignature string) []byte {
+	if !functionSignatureRegex.MatchString(functionSignature) {
+		panic(fmt.Errorf("invalid function signature: %q", functionSignature))
+	}
+	hash := crypto.Keccak256([]byte(functionSignature))
+	return hash[:4]
+}
+
+// DeductGas checks if [suppliedGas] is sufficient against [requiredGas] and deducts [requiredGas] from [suppliedGas].
+func DeductGas(suppliedGas uint64, requiredGas uint64) (uint64, error) {
```
Comment (nddeluca) [inline precompile/contract/utils.go:32]: Missing test coverage

File: precompile/modules/registerer.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	registeredModules = make([]Module, 0)
+)
+
+// RegisterModule registers a stateful precompile module
+func RegisterModule(stm Module) error {
+	address := stm.Address
+
+	for _, registeredModule := range registeredModules {
+		if registeredModule.Address == address {
+			return fmt.Errorf("address %s already used by a stateful precompile", address)
```
Comment (nddeluca) [inline precompile/modules/registerer.go:25]: Not sure this error case is tested

File: core/vm/evm.go [hunk 6]
```diff
@@ -51,8 +53,20 @@ func (evm *EVM) precompile(addr common.Address) (PrecompiledContract, bool) {
 	default:
 		precompiles = PrecompiledContractsHomestead
 	}
+
+	// Check the existing precompiles first
```
Comment (nddeluca) [inline core/vm/evm.go:57]: This is important to behavior to test -- let's add a test that overrides an existing precompile address with a contract that panics and test that it does not override it.

File: core/vm/evm.go [hunk 7]
```diff
@@ -381,7 +400,7 @@ func (evm *EVM) StaticCall(caller ContractRef, addr common.Address, input []byte
 	}
 
 	if p, isPrecompile := evm.precompile(addr); isPrecompile {
-		ret, gas, err = RunPrecompiledContract(p, input, gas)
+		ret, gas, err = RunStatefulPrecompiledContract(p, evm, caller.Address(), addr, input, gas, evm.interpreter.readOnly)
```
Comment (nddeluca) [inline core/vm/evm.go]: I'm not 100% we should be passing readOnly var for static call -- is this always true or is it only true run in non-tx contexts?  Need to ensure an executed precompile is able to infer & panic if tx requests a state changing call in staticcall

File: precompile/contract/contract.go (hunk 1 above)
Comment (evgeniy-scherbina) [reply precompile/contract/contract.go:60]: Fixed https://github.com/Kava-Labs/go-ethereum/pull/10/commits/a7051557696ec1b5debe28fdcaf0b19138781b40

File: precompile/contract/utils.go (hunk 2 above)
Comment (evgeniy-scherbina) [reply precompile/contract/utils.go]: Fixed https://github.com/Kava-Labs/go-ethereum/pull/10/commits/22e3b2347bfce0e4c32380c4939f64ecc89fc54e

File: precompile/contract/contract.go (hunk 3 above)
Comment (evgeniy-scherbina) [reply precompile/contract/contract.go:82]: @nddeluca it's tested here: https://github.com/Kava-Labs/go-ethereum/blob/yevhenii/stateful-precompile-implementation/precompile/contracts/sum3/contract_test.go#L211-L252 inside `sum3` smart contract (in another package).

Is it enough? or you think it make sense to add tests in the same `contract` package?

Comment (evgeniy-scherbina) [reply precompile/contract/contract.go:82]: @nddeluca Fixed https://github.com/Kava-Labs/go-ethereum/pull/10/commits/02c413978db9150f5b4f4226be935de480fd2d1d

I think it make sense to have such tests in `contract` package as well, in the future we can decide to remove `sum3` precompile

File: precompile/contract/utils.go (hunk 4 above)
Comment (evgeniy-scherbina) [reply precompile/contract/utils.go:32]: Fixed https://github.com/Kava-Labs/go-ethereum/pull/10/commits/0de25b9d9966182024f21721ceedb2db6d15a09d

File: precompile/modules/registerer.go (hunk 5 above)
Comment (evgeniy-scherbina) [reply precompile/modules/registerer.go:25]: Fixed https://github.com/Kava-Labs/go-ethereum/pull/10/commits/671ae7bed3d5d22a89e052acb053c9968982c921

File: core/vm/evm.go (hunk 7 above)
Comment (evgeniy-scherbina) [reply core/vm/evm.go]: Fixed:
- https://github.com/Kava-Labs/go-ethereum/pull/10/commits/74993c4c1ed5f7ee9edbfb4513b6707a471f752b
- https://github.com/Kava-Labs/go-ethereum/pull/10/commits/e08309004de5e54f20c67549ada19f6133a66706
- https://github.com/Kava-Labs/go-ethereum/pull/10/commits/57724b99e87b5072a28e07fcd18cca97ff8e5d5b

File: core/vm/evm.go (hunk 6 above)
Comment (evgeniy-scherbina) [reply core/vm/evm.go:57]: Fixed https://github.com/Kava-Labs/go-ethereum/pull/10/commits/d85296e76ae9da578bb199d2a82207157263623b

In `TestEvmIsPrecompileMethod` test, I'm testing only `stateful precompile` part, but we can consider adding tests for `native stateless precompile` as well if you think it make sense.

File: contracts/contracts/ExampleSum3.sol
Comment (evgeniy-scherbina) [reply contracts/contracts/ExampleSum3.sol:1]: Fixed https://github.com/Kava-Labs/go-ethereum/pull/10/commits/2a7165800395c1777d32e9cda92ae316cd6f9c53

**Review by nddeluca — APPROVED** [review]

**Pull request merged**